test/unit/org/apache/cassandra/config/ConfigCompatabilityTest.java

```diff
@@ -0,0 +1,360 @@
+        {
+            Property property = e.getValue();
+            Map<String, Property> subProperties = Properties.isPrimitive(property) || Properties.isCollection(property) ? Collections.emptyMap() : loader.getProperties(property.getType());
+            Node subNode;
+            if (subProperties.isEmpty())
+            {
+                subNode = new Leaf(property.getType().getName());
+            }
+            else
+            {
+                Klass subKlass = new Klass(property.getType());
+                addProperties(loader, subKlass, property.getType());
+                subNode = subKlass;
+            }
+            node.addProprty(e.getKey(), subNode);
+        }
+    }
+
+    @JsonSerialize(using = NodeSerializer.class)
+    @JsonDeserialize(using = NodeDeserializer.class)
+    private interface Node
+    {
+
+    }
+
+    private static final class NodeSerializer extends StdSerializer<Node>
+    {
+        NodeSerializer()
+        {
+            super(Node.class);
+        }
+
+        @Override
+        public void serialize(Node node, JsonGenerator jsonGenerator, SerializerProvider serializerProvider) throws IOException
+        {
+            if (node instanceof Leaf)
+            {
+                jsonGenerator.writeString(((Leaf) node).type);
+            }
+            else if (node instanceof Klass)
+            {
+                Klass klass = (Klass) node;
+                jsonGenerator.writeStartObject();
+                for (Map.Entry<String, Node> e : klass.properties.entrySet())
+                    jsonGenerator.writeObjectField(e.getKey(), e.getValue());
+                jsonGenerator.writeEndObject();
+            }
+        }
+    }
+
+    private static final class NodeDeserializer extends StdDeserializer<Node>
+    {
+
+        protected NodeDeserializer()
+        {
+            super(Node.class);
+        }
+
+        @Override
+        public Node deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException, JsonProcessingException
+        {
+            return toNode(jsonParser.getCodec().readTree(jsonParser));
+        }
+
+        private static Node toNode(TreeNode node)
+        {
+            if (node.isValueNode())
+                return new Leaf(((TextNode) node).textValue());
+            Map<String, Node> props = new HashMap<>();
+            Iterator<String> it = node.fieldNames();
+            while (it.hasNext())
+            {
+                String name = it.next();
+                Node value = toNode(node.get(name));
+                Node previous = props.put(name, value);
+                if (previous != null)
+                    throw new AssertionError("Duplicate properties found: " + name);
+            }
+            Klass klass = new Klass();
+            klass.setProperties(props);
+            return klass;
+        }
+    }
+
+    private static class Klass implements Node
```

Comment:
nit: Not super important, but perhaps we could name this like `ClassTree` or something?